cpp/src/lists/interleave_columns.cu

```diff
@@ -0,0 +1,359 @@
+                               mr);
+  }
+
+  template <class T>
+  std::enable_if_t<cudf::is_fixed_width<T>(), std::unique_ptr<column>> operator()(
+    table_view const& input,
+    column_view const& output_list_offsets,
+    size_type num_output_lists,
+    size_type num_output_entries,
+    bool has_null_mask,
+    rmm::cuda_stream_view stream,
+    rmm::mr::device_memory_resource* mr) const noexcept
+  {
+    auto const num_cols     = input.num_columns();
+    auto const num_rows     = input.num_rows();
+    auto const table_dv_ptr = table_device_view::create(input);
+
+    // The output child column.
+    auto const child_col = lists_column_view(*input.begin()).child();
+    auto output =
+      allocate_like(child_col, num_output_entries, mask_allocation_policy::NEVER, stream, mr);
+    auto output_dv_ptr = mutable_column_device_view::create(*output);
+
+    // The array of int8_t to store entry validities.
+    auto validities = rmm::device_uvector<int8_t>(has_null_mask ? num_output_entries : 0, stream);
+
+    thrust::for_each_n(
+      rmm::exec_policy(stream),
+      thrust::make_counting_iterator<size_type>(0),
+      num_output_lists,
+      [num_cols,
+       table_dv         = *table_dv_ptr,
+       d_validities     = validities.begin(),
+       dst_list_offsets = output_list_offsets.begin<offset_type>(),
+       d_output         = output_dv_ptr->begin<T>(),
+       has_null_mask] __device__(size_type const dst_list_id) {
+        auto const src_col_id     = dst_list_id % num_cols;
+        auto const src_list_id    = dst_list_id / num_cols;
+        auto const& src_lists_col = table_dv.column(src_col_id);
+        auto const src_list_offsets =
+          src_lists_col.child(lists_column_view::offsets_column_index).data<size_type>() +
+          src_lists_col.offset();
+        auto const& src_child = src_lists_col.child(lists_column_view::child_column_index);
+
+        // The indices of the entries in the list to gather from.
+        auto const start_idx = src_list_offsets[src_list_id];
+        auto const end_idx   = src_list_offsets[src_list_id + 1];
+
+        // Fill the validities array if necessary.
+        if (has_null_mask) {
+          for (auto read_idx  = src_list_offsets[src_list_id],
+                    end_idx   = src_list_offsets[src_list_id + 1],
+                    write_idx = dst_list_offsets[dst_list_id];
+               read_idx < end_idx;
+               ++read_idx, ++write_idx) {
+            auto const is_valid     = src_child.is_valid(read_idx);
+            d_validities[write_idx] = static_cast<int8_t>(is_valid);
+          }
+        }
+
+        // Do a memcpy for the entire list entries.
+        auto const input_ptr  = reinterpret_cast<char const*>(src_child.data<T>() + start_idx);
+        auto const output_ptr = reinterpret_cast<char*>(&d_output[dst_list_offsets[dst_list_id]]);
+        thrust::copy(
+          thrust::seq, input_ptr, input_ptr + sizeof(T) * (end_idx - start_idx), output_ptr);
```

Comment:
```suggestion
        // Do a memcpy for the entire list entries.
        using DeviceT    = device_storage_type_t<T>;        
        auto const input_ptr  = reinterpret_cast<char const*>(src_child.data<DeviceT>() + start_idx);
        auto const output_ptr = reinterpret_cast<char*>(&d_output[dst_list_offsets[dst_list_id]]);
        thrust::copy(
          thrust::seq, input_ptr, input_ptr + sizeof(DeviceT) * (end_idx - start_idx), output_ptr);
```

Need to use `device_storage_type_t` to handle fixed point types correctly.   Should add some fixed point tests below as well.

Reply:
We don't need that, as `T` is already a `dispatch_storage_type`.